tracing sample editorial review (#5000)
Particular/docs.particular.net: samples/tracing/sample.md

---
title: Distributed tracing with Azure Monitor Application Insights
summary: How to extend the NServiceBus pipeline to send trace data to Azure Monitor Application Insights
component: Core
reviewed: 2020-09-10
related:
 - monitoring/metrics
---

## Introduction

.NET Core 3.1 introduced support for distributed tracing. Types in the [System.Diagnostics](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics?view=dotnet-plat-ext-3.1) namespace may be used to expose and correlate execution data from a distributed system. This sample shows how to extend the NServiceBus pipeline with custom behaviors that publish trace data and how to send that data to [Azure Monitor Application Insights](https://docs.microsoft.com/en-us/azure/azure-monitor/app/app-insights-overview) for analysis and visualization.

## Running the sample

Running the sample requires an Application Insights resource. The [Instrumentation Key](https://docs.microsoft.com/en-us/azure/bot-service/bot-service-resources-app-insights-keys?view=azure-bot-service-4.0#instrumentation-key) must be provided by setting the `ApplicationInsightsKey` environment variable.

> After starting the sample, pressing <kbd>Enter</kbd> sends an `InitialCommand`. When that command is handled, a `FollowUpEvent` is published. Trace data for both actions is pushed to Application Insights and available as an [end-to-end transaction](https://docs.microsoft.com/en-us/azure/azure-monitor/app/transaction-diagnostics#transaction-diagnostics-experience) trace:

Author: > [end-to-end transaction](https://docs.microsoft.com/en-us/azure/azure-monitor/app/transaction-diagnostics#transaction-diagnostics-experience) trace

I don't see any mention of "end-to-end" in the linked page.

Reviewer: Good catch. I've added suggestion. It used to be called end-to-end transactions - you can still see that in at the top of the screenshot that the link points to.

> After starting the sample, pressing <kbd>Enter</kbd> sends an `InitialCommand`. When that command is handled, a `FollowUpEvent` is published. Trace data for both actions is pushed to Application Insights and available as an [end-to-end transaction](https://docs.microsoft.com/en-us/azure/azure-monitor/app/transaction-diagnostics#transaction-diagnostics-experience) trace:

Reviewer: ```suggestion
After starting the sample, pressing <kbd>Enter</kbd> sends an `InitialCommand`. When that command is handled, a `FollowUpEvent` is published. Trace data for both actions is pushed to Application Insights and available on the [transaction diagnostics](https://docs.microsoft.com/en-us/azure/azure-monitor/app/transaction-diagnostics#transaction-diagnostics-experience) pane:
```
